lib/llhttp/parser.js

```diff
@@ -0,0 +1,241 @@
+'use strict'
+
+/* global WebAssembly */
+
+const { resolve } = require('path')
+const { readFileSync } = require('fs')
+const constants = require('./constants')
+const { kMaxHeadersSize } = require('../core/symbols')
+const WASM_BUILD = resolve(__dirname, './llhttp.wasm')
+const bin = readFileSync(WASM_BUILD)
+const mod = new WebAssembly.Module(bin)
+
+const kOnMessageBegin = 0
+const kOnHeaders = 1
+const kOnHeadersComplete = 2
+const kOnBody = 3
+const kOnMessageComplete = 4
+const kOnExecute = 5
+
+const kPtr = Symbol('kPrt')
+const kStatusMessage = Symbol('kStatusMessage')
+const kHeadersFields = Symbol('kHeadersFields')
+const kHeadersValues = Symbol('kHeadersValues')
+const kBufferSize = Symbol('kBufferSize')
+const kHeadersTimeout = Symbol('kHeadersTimeout')
+const kBufferPtr = Symbol('kBufferPtr')
+const kMallocBuffer = Symbol('kMallocBuffer')
+const kCurrentBuffer = Symbol('kCurrentBuffer')
+const kGetHeaders = Symbol('kGetHeaders')
+const kMakeError = Symbol('kMakeError')
+
+/**
+ * Current parser reference
+ */
+let currentParser = null
+
+const cstr = (ptr, len) =>
+  Buffer.from(inst.exports.memory.buffer, ptr, len).toString()
+
+/* eslint-disable camelcase */
+const wasm_on_message_begin = p => {
+  currentParser[kStatusMessage] = null
+  return currentParser[kOnMessageBegin]()
+}
+
+const wasm_on_url = (p, at, length) => {
+  return 0
+}
+
+const wasm_on_status = (p, at, length) => {
+  currentParser[kStatusMessage] = cstr(at, length)
+  return 0
+}
+
+const wasm_on_header_field = (p, at, length) => {
+  currentParser[kHeadersFields].push(cstr(at, length))
+  return 0
+}
+
+const wasm_on_header_value = (p, at, length) => {
+  currentParser[kHeadersValues].push(cstr(at, length))
+  return 0
+}
```

Comment:
We could maybe make this faster:

```js
const wasm_on_header_field = (p, at, length) => {
  const idx = currentParser[kHeaderFieldsCount]++ * 2 + 0
  if (currentParser[kRawHeaders].length < idx) {
    currentParser[kRawHeaders].length = idx + 16
  }
  currentParser[kRawHeaders][idx] = cstr(at, length)
  return 0
}

const wasm_on_header_value = (p, at, length) => {
  const idx = currentParser[kHeaderValuesCount]++ * 2 + 1
  if (currentParser[kRawHeaders].length < idx) {
    currentParser[kRawHeaders].length = idx + 16
  }
  currentParser[kRawHeaders][idx] = cstr(at, length)
  return 0
}

[kGetHeaders] () {
  const rawHeaders = currentParser[kRawHeaders]
  rawHeaders.length = currentParser[kHeaderValuesCount] * 2
  currentParser[kRawHeaders] = new Array(16)
  
  this[kHeaderFieldsCount] = 0
  this[kHeaderValuesCount] = 0

  return rawHeaders
}
```

But yea, maybe follow up PR's :D

Reply:
Should I add a note?